python/_tskitmodule.c

```diff
@@ -6073,6 +6073,86 @@ TableCollection_get_file_uuid(TableCollection *self, void *closure)
     return Py_BuildValue("s", self->tables->file_uuid);
 }
 
+static PyObject *
+TableCollection_get_metadata(TableCollection *self, void *closure)
+{
+    PyObject *ret = NULL;
+
+    /* Py_BuildValue returns Py_None for zero length, we would rather
+       return a zero-length string */
+    if (self->tables->metadata_length == 0) {
+        ret = PyBytes_FromString("");
+    } else {
+        ret = Py_BuildValue("y#", self->tables->metadata, self->tables->metadata_length);
+    }
+    return ret;
+}
+
+static int
+TableCollection_set_metadata(TableCollection *self, PyObject *arg, void *closure)
+{
+    int ret = -1;
+    int err;
+    char *metadata;
+    Py_ssize_t metadata_length;
+
+    if (arg == NULL) {
+        PyErr_Format(
+            PyExc_AttributeError,
+            "Cannot del metadata, set to empty string (b\"\") to clear.");
+        goto out;
+    }
+    err = PyBytes_AsStringAndSize(arg, &metadata, &metadata_length);
+    if (err != 0) {
+        goto out;
+    }
+    err = tsk_table_collection_set_metadata(
+        self->tables, metadata, metadata_length);
+    if (err != 0) {
+        handle_library_error(err);
+        goto out;
+    }
+    ret = 0;
+out:
+    return ret;
+}
+
+static PyObject *
+TableCollection_get_metadata_schema(TableCollection *self, void *closure)
+{
+    return make_Py_Unicode_FromStringAndLength(
+        self->tables->metadata_schema, self->tables->metadata_schema_length);
+}
+
+static int
+TableCollection_set_metadata_schema(TableCollection *self, PyObject *arg, void *closure)
+{
+    int ret = -1;
+    int err;
+    const char *metadata_schema;
+    Py_ssize_t metadata_schema_length;
+
+    if (arg == NULL) {
```

Comment:
Same here - we're repeating the same code a good few times.

Reply:
As above.